<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Chap9-PractiseSet</title>
</head>
<body>
    <script>
        // ---------------------------- PROBLEM 01 --------------------------------
        const LoadScript=(src)=>{
            return new Promise((resolve,reject)=>{
                let script = document.createElement("script")
                script.src = src
                document.head.append(script)
                script.onload=()=>{
                    resolve("Script 1 done")
                }
                script.onerror=()=>{
                    reject(0)
                }
            })
        }

        let p1 = LoadScript("https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/js/bootstrap.bundle.min.js")
        p1.then((value)=>{
            console.log(value)
            console.log("Script 1.then done")
        })

        // ---------------------------- PROBLEM 02 --------------------------------
        const loadjs = async (src) => {
        return new Promise((resolve, reject) => {
                let script = document.createElement("script")
                script.src = src
                script.onload = () => {
                        resolve(" Done success")
                }
                document.head.append(script)
        })
        }

        const main2 = async () => {
           console.log(new Date().getMilliseconds())
           let a = await loadjs("https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js")
           console.log(a)
           console.log(new Date().getMilliseconds())
        }
        main2()

        // ---------------------------- PROBLEM 03 --------------------------------
        let p = () => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                        reject(new Error("Please this is not acceptable"))
                }, 3000)
            })
        }

        let a = async () => {
            try {
                let c = await p()
                console.log(c)
            }
            catch (err) {
                console.log("Problem 3 error has been handled")
            }
        }
        a()
        
        // ---------------------------- PROBLEM 04 --------------------------------
        let pm1 = async () => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                        resolve(10)
                }, 2000)
            })
        }
        let pm2 = async () => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                        resolve(20)
                }, 1000)
            })
        }
        let pm3 = async () => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                        resolve(30)
                }, 3000)
            })
        }

        const run = async () => {
            console.time("run")
            // let a1 = await p1()      // Fetch first 10 products from the database
            // let a2 = await p2()      // Fetch another 10 products from the database
            // let a3 = await p3()      // Fetch yet another 10 products from the database


            let a1 = pm1()               // Fetch first 10 products from the database
            let a2 = pm2()               // Fetch another 10 products from the database
            let a3 = pm3()               // Fetch yet another 10 products from the database
            let a1a2a3 = await Promise.all([a1, a2, a3])
            console.log(a1a2a3)
            // console.log(a1, a2, a3)
            console.timeEnd("run")
        }
        run()

    </script>
</body>
</html>